### DISCUSSION TITLE
fitting free water mapping out of iterations

### DISCUSSION THREAD

**Discussion by HusBunia** [dipy/dipy#3577] [Reconstruction] [answered]
## Description
Dear developer:

I having using the fwdtimodel and fwdtifit to map the free water compartment. 
  
The test data is download from HCP, with 288 directions and 2 b-values:
The code is as follows: 

```python
   fwdtimodel = fwdti.FreeWaterTensorModel(gtab)
    fwdtifit = fwdtimodel.fit(data, mask=mask)
    fwvolume = fwdtifit.f
```

But there are some warning information: 
```python
/usr/local/lib64/python3.6/site-packages/scipy/optimize/minpack.py:475: RuntimeWarning: Number of calls to function has reached maxfev = 1800.
 warnings.warn(errors[info][0], RuntimeWarning)
/usr/local/lib64/python3.6/site-packages/dipy/reconst/fwdti.py:311: RuntimeWarning: overflow encountered in exp
  SIpred = (1-FS)*np.exp(np.dot(W, all_new_params)) + FS*S0*SFW.T

/usr/local/lib64/python3.6/site-packages/dipy/reconst/fwdti.py:312: RuntimeWarning: overflow encountered in square
  F2 = np.sum(np.square(SI - SIpred), axis=0)

/usr/local/lib64/python3.6/site-packages/dipy/reconst/fwdti.py:458: RuntimeWarning: overflow encountered in exp
   y = (1-f) * np.exp(np.dot(design_matrix, tensor[:7])) + \
```

I wonder whether these warnings have influenced the output images and anyway to fix them?

 Many thanks for your help!

**Comment by skoudoro** [comment, marked as answer]
Hi @HusBunia,

Usually, this kind of warning appears in the background of your data.  What is your preprocessing step?  Do you apply a mask? Do you denoise your data?

It would be good to check first which voxel generate those warning and if it is the background, no worries otherwise, we will look deeper.

Let us know when you know more and feel free to describe your preprocessing step.

**Comment by HusBunia** [comment]
> Hi @HusBunia,
> 
> Usually, this kind of warning appears in the background of your data. What is your preprocessing step? Do you apply a mask? Do you denoise your data?
> 
> It would be good to check first which voxel generate those warning and if it is the background, no worries otherwise, we will look deeper.
> 
> Let us know when you know more and feel free to describe your preprocessing step.

Hi @skoudoro 

Many thanks for your reply! 

The data I used is download from HCP. A standard HCP preprocessed pipeline has performed on the data, and the whole brain mask was used on the image. Follow your suggestion, I found the data with the perfect brain mask did not appear the warning. I guess the warning appeared when the noise non-brain tissues are included. I have only one last question. I found a significant dark band appeared in the boundary between different tissues, such as gray matter and white matter.  A similar dark band has also been found in the tutorial example. Could you please tell me what caused the dark band to appear?

Thanks!

**Comment by skoudoro** [comment]
Hi @HusBunia,

Great! glad to see that the warning was only in a non-brain tissue as expected. Thank you for your feedback.

>  found a significant dark band appeared in the boundary between different tissues, such as gray matter and white matter. A similar dark band has also been found in the tutorial example. Could you please tell me what caused the dark band to appear?
 
Concerning this question, I will let @ShreyasFadnavis, @RafaelNH, @arokem, @mvgolub, or @alecrimi answer you. One of them will provide you a more detailed answer.

Cheers !

**Comment by HusBunia** [comment]
> Hi @HusBunia,
> 
> Great! glad to see that the warning was only in a non-brain tissue as expected. Thank you for your feedback.
> 
> > found a significant dark band appeared in the boundary between different tissues, such as gray matter and white matter. A similar dark band has also been found in the tutorial example. Could you please tell me what caused the dark band to appear?
> 
> Concerning this question, I will let @ShreyasFadnavis, @RafaelNH, @arokem, @mvgolub, or @alecrimi answer you. One of them will provide you a more detailed answer.
> 
> Cheers !

Hi @skoudoro
Many thanks for your reply!  Here is a freewater map reconstructed with dipy's fwdti module.  The dark band has labeled by a red rectangle.  Data comes from HCP, and the b value larger than 2000 have been excluded.  
Many thanks for your help again!

![fwvolume2](https://user-images.githubusercontent.com/40214939/122172179-17481580-ceb3-11eb-924a-e3bef0370903.png)
